File: app/models/period.rb
class Period
  TYPES = %i[weekly monthly yearly].freeze

  class InvalidParam < ArgumentError; end

  class << self
    def latest_closed(type, now: Time.current)
      local = now.in_time_zone(Time.zone)
      case type.to_sym
      when :weekly  then new(type: :weekly,  starts_at: local.beginning_of_week - 7.days)
      when :monthly then new(type: :monthly, starts_at: local.beginning_of_month.prev_month)
      when :yearly  then new(type: :yearly,  starts_at: local.beginning_of_year.prev_year)
      else raise ArgumentError, "unknown period type: #{type.inspect}"
      end
    end

    def parse(type, param)
      case type.to_sym
      when :weekly
        m = param.to_s.match(/\A(\d{4})-W(\d{1,2})\z/) or raise InvalidParam, "invalid weekly param: #{param.inspect}"
        date = Date.commercial(m[1].to_i, m[2].to_i, 1)
        new(type: :weekly, starts_at: Time.zone.local(date.year, date.month, date.day))
      when :monthly
        m = param.to_s.match(/\A(\d{4})-(\d{1,2})\z/) or raise InvalidParam, "invalid monthly param: #{param.inspect}"
        new(type: :monthly, starts_at: Time.zone.local(m[1].to_i, m[2].to_i, 1))
      when :yearly
        m = param.to_s.match(/\A(\d{4})\z/) or raise InvalidParam, "invalid yearly param: #{param.inspect}"
        new(type: :yearly, starts_at: Time.zone.local(m[1].to_i, 1, 1))
      else
        raise ArgumentError, "unknown period type: #{type.inspect}"
      end
    rescue Date::Error => e
      raise InvalidParam, e.message
    end
  end

  attr_reader :type, :starts_at

  def initialize(type:, starts_at:)
    @type = type.to_sym
    raise ArgumentError, "unknown period type: #{@type}" unless TYPES.include?(@type)
    @starts_at = normalize(starts_at)
  end

  def ends_at
    case type
    when :weekly  then starts_at + 7.days
    when :monthly then starts_at + 1.month
    when :yearly  then starts_at + 1.year
    end
  end

  def prev
    self.class.new(type: type, starts_at: step(-1))
  end

  def next
    self.class.new(type: type, starts_at: step(+1))
  end

  def complete?(now: Time.current)
    ends_at <= now
  end

  def contains?(time)
    t = time.in_time_zone(Time.zone)
    starts_at <= t && t < ends_at
  end

  def label
    case type
    when :weekly  then "#{starts_at.strftime('%B %-d')} — #{(ends_at - 1.day).strftime('%B %-d')}"
    when :monthly then starts_at.strftime("%B %Y")
    when :yearly  then starts_at.strftime("%Y")
    end
  end

  def subtitle
    case type
    when :weekly
      date = starts_at.to_date
      "Week #{date.cweek} · #{date.cwyear}"
    else
      nil
    end
  end

  def to_param
    case type
    when :weekly
      date = starts_at.to_date
      format("%d-W%02d", date.cwyear, date.cweek)
    when :monthly then starts_at.strftime("%Y-%m")
    when :yearly  then starts_at.strftime("%Y")
    end
  end

  def ==(other)
    other.is_a?(Period) && other.type == type && other.starts_at == starts_at
  end
  alias_method :eql?, :==

  def hash
    [type, starts_at].hash
  end

  private

  def step(direction)
    case type
    when :weekly  then starts_at + direction * 7.days
    when :monthly then direction.positive? ? starts_at.next_month : starts_at.prev_month
    when :yearly  then direction.positive? ? starts_at.next_year  : starts_at.prev_year
    end
  end

  def normalize(time)
    ts = time.in_time_zone(Time.zone)
    case type
    when :weekly  then ts.beginning_of_week
    when :monthly then ts.beginning_of_month
    when :yearly  then ts.beginning_of_year
    end
  end
end

File: src/period.ts
import {
  addMonths,
  addWeeks,
  addYears,
  getISOWeek,
  getISOWeekYear,
  startOfMonth,
  startOfWeek,
  startOfYear,
  subDays,
} from "date-fns";
import { formatInTimeZone, fromZonedTime, toZonedTime } from "date-fns-tz";

export type PeriodType = "weekly" | "monthly" | "yearly";

export const PERIOD_TYPES = ["weekly", "monthly", "yearly"] as const satisfies readonly PeriodType[];

const STEPPERS: Record<PeriodType, (d: Date, n: number) => Date> = {
  weekly: addWeeks,
  monthly: addMonths,
  yearly: addYears,
};

const STARTERS: Record<PeriodType, (d: Date) => Date> = {
  weekly: (d) => startOfWeek(d, { weekStartsOn: 1 }),
  monthly: startOfMonth,
  yearly: startOfYear,
};

export class InvalidParam extends Error {
  constructor(message: string) {
    super(message);
    this.name = "InvalidParam";
  }
}

export class Period {
  readonly type: PeriodType;
  readonly startsAt: Date;
  readonly timeZone: string;

  private _localStart?: Date;
  private _endsAt?: Date;
  private _param?: string;
  private _label?: string;
  private _subtitle: string | null | undefined = undefined;

  constructor(opts: { type: PeriodType; startsAt: Date; timeZone: string }) {
    if (!PERIOD_TYPES.includes(opts.type)) {
      throw new TypeError(`unknown period type: ${opts.type as string}`);
    }
    this.type = opts.type;
    this.timeZone = opts.timeZone;
    const local = toZonedTime(opts.startsAt, opts.timeZone);
    this.startsAt = fromZonedTime(STARTERS[opts.type](local), opts.timeZone);
  }

  static latestClosed(type: PeriodType, now: Date, timeZone: string): Period {
    const local = toZonedTime(now, timeZone);
    const prev = STEPPERS[type](STARTERS[type](local), -1);
    return new Period({ type, startsAt: fromZonedTime(prev, timeZone), timeZone });
  }

  static parse(type: PeriodType, param: string, timeZone: string): Period {
    return new Period({
      type,
      startsAt: fromZonedTime(parseWallStart(type, param), timeZone),
      timeZone,
    });
  }

  get endsAt(): Date {
    return (this._endsAt ??= fromZonedTime(STEPPERS[this.type](this.localStart, 1), this.timeZone));
  }

  prev(): Period {
    return this.step(-1);
  }

  next(): Period {
    return this.step(1);
  }

  isComplete(now: Date): boolean {
    return this.endsAt.getTime() <= now.getTime();
  }

  contains(time: Date): boolean {
    const t = time.getTime();
    return this.startsAt.getTime() <= t && t < this.endsAt.getTime();
  }

  get label(): string {
    return (this._label ??= this.computeLabel());
  }

  get subtitle(): string | null {
    if (this._subtitle === undefined) {
      this._subtitle = this.type === "weekly"
        ? `Week ${getISOWeek(this.localStart)} · ${getISOWeekYear(this.localStart)}`
        : null;
    }
    return this._subtitle;
  }

  get param(): string {
    return (this._param ??= this.computeParam());
  }

  equals(other: Period): boolean {
    return (
      this.type === other.type &&
      this.startsAt.getTime() === other.startsAt.getTime() &&
      this.timeZone === other.timeZone
    );
  }

  private get localStart(): Date {
    return (this._localStart ??= toZonedTime(this.startsAt, this.timeZone));
  }

  private fmt(d: Date, format: string): string {
    return formatInTimeZone(d, this.timeZone, format);
  }

  private computeLabel(): string {
    switch (this.type) {
      case "weekly": {
        const start = this.fmt(this.startsAt, "MMMM d");
        const end = this.fmt(subDays(this.endsAt, 1), "MMMM d");
        return `${start} — ${end}`;
      }
      case "monthly":
        return this.fmt(this.startsAt, "MMMM yyyy");
      case "yearly":
        return this.fmt(this.startsAt, "yyyy");
    }
  }

  private computeParam(): string {
    switch (this.type) {
      case "weekly":
        return `${getISOWeekYear(this.localStart)}-W${String(getISOWeek(this.localStart)).padStart(2, "0")}`;
      case "monthly":
        return this.fmt(this.startsAt, "yyyy-MM");
      case "yearly":
        return this.fmt(this.startsAt, "yyyy");
    }
  }

  private step(direction: -1 | 1): Period {
    const stepped = STEPPERS[this.type](this.localStart, direction);
    return new Period({
      type: this.type,
      startsAt: fromZonedTime(stepped, this.timeZone),
      timeZone: this.timeZone,
    });
  }
}

function parseWallStart(type: PeriodType, param: string): Date {
  switch (type) {
    case "weekly": {
      const m = param.match(/^(\d{4})-W(\d{1,2})$/);
      if (!m) throw new InvalidParam(`invalid weekly param: ${param}`);
      const date = isoCommercialDate(parseInt(m[1]!, 10), parseInt(m[2]!, 10));
      if (!date) throw new InvalidParam(`invalid weekly param: ${param}`);
      return date;
    }
    case "monthly": {
      const m = param.match(/^(\d{4})-(\d{1,2})$/);
      if (!m) throw new InvalidParam(`invalid monthly param: ${param}`);
      const month = parseInt(m[2]!, 10);
      if (month < 1 || month > 12) throw new InvalidParam(`invalid monthly param: ${param}`);
      return new Date(parseInt(m[1]!, 10), month - 1, 1);
    }
    case "yearly": {
      const m = param.match(/^(\d{4})$/);
      if (!m) throw new InvalidParam(`invalid yearly param: ${param}`);
      return new Date(parseInt(m[1]!, 10), 0, 1);
    }
  }
}

function isoCommercialDate(year: number, week: number): Date | null {
  if (week < 1 || week > 53) return null;
  const jan4 = new Date(year, 0, 4);
  const dow = jan4.getDay() === 0 ? 7 : jan4.getDay();
  const week1Mon = new Date(year, 0, 4 - dow + 1);
  const target = addWeeks(week1Mon, week - 1);
  if (getISOWeekYear(target) !== year || getISOWeek(target) !== week) return null;
  return target;
}
